### PR TITLE
[CARBONDATA-2420][32K] Support string longer than 32000 characters

### PR DESCRIPTION
**Pull request by xuchuanyin**
Add a property in creating table 'long_string_columns' to support string columns that will contains more than 32000 characters.
Inside carbondata, it use an integer instead of short to store the length of bytes content.

Internally in Carbondata,
1. add a new datatype called `varchar` to represent the long string column
2. add a new encoding called `DIRECT_COMPRESS_VARCHAR` to the varcher column page meta
3. use an integer (previously short) to store the length of bytes content.
4. add 2GB constraint for one column page

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [x] Any interfaces changed?
 `Only internal interfaces have been changed`
 - [x] Any backward compatibility impacted?
 `NO`
 - [x] Document update required?
`YES`
 - [x] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
`Added tests`
        - How it is tested? Please attach test report.
`Tested in local machine`
        - Is it a performance related change? Please attach the performance test report.
`NO`
        - Any additional information to help reviewers in testing this change.
      `NA` 
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
`NA`

### CODE REVIEW

**Comment by xuchuanyin** [comment]
@kumarvishal09 
Can you review and merge it? The previous discussion can be found in #2252

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6366/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5204/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5315/

**Comment by xuchuanyin** [comment]
retest this please

**Comment by xuchuanyin** [comment]
retest it please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6372/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5210/

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 1]
```diff
@@ -1601,6 +1602,8 @@
   // As Short data type is used for storing the length of a column during data processing hence
   // the maximum characters that can be supported should be less than Short max value
   public static final int MAX_CHARS_PER_COLUMN_DEFAULT = 32000;
+  // todo: use infinity first, will switch later
+  public static final int MAX_CHARS_PER_COLUMN_INFINITY = -1;
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: why this is -1?? as per previous discussion we can support max 67104 chars

File: core/src/main/java/org/apache/carbondata/core/datastore/page/VarLengthColumnPageBase.java [hunk 2]
```diff
@@ -289,6 +289,12 @@ public void putDouble(int rowId, double value) {
 
   @Override
   public void putBytes(int rowId, byte[] bytes) {
+    // rowId * 4 represents the length of L in LV
+    if (bytes.length > (Integer.MAX_VALUE - totalLength - rowId * 4)) {
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/datastore/page/VarLengthColumnPageBase.java]: Better to add a check in FieldEncoder to validate if it crosses 67104 bytes for varchar datatype, it should be fail first mechanism, here it will fail while writing which is too late as most of the data transformation is done.

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RawColumnChunkUtil.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
+/*
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RawColumnChunkUtil.java]: Is it related to this PR, if not please raise in separate PR

File: core/src/main/java/org/apache/carbondata/core/datastore/page/encoding/DefaultEncodingFactory.java [hunk 4]
```diff
@@ -103,6 +103,7 @@ private ColumnPageEncoder createEncoderForDimensionLegacy(TableSpec.DimensionSpe
         return new HighCardDictDimensionIndexCodec(
             dimensionSpec.isInSortColumns(),
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/datastore/page/encoding/DefaultEncodingFactory.java]: Instead of passing all the properties of dimensionspec individually, pass dimensionspec it self.

File: core/src/main/java/org/apache/carbondata/core/datastore/page/encoding/EncodingFactory.java [hunk 5]
```diff
@@ -71,7 +72,7 @@ public ColumnPageDecoder createDecoder(List<Encoding> encodings, List<ByteBuffer
     byte[] encoderMeta = encoderMetas.get(0).array();
     ByteArrayInputStream stream = new ByteArrayInputStream(encoderMeta);
     DataInputStream in = new DataInputStream(stream);
-    if (encoding == DIRECT_COMPRESS) {
+    if (encoding == DIRECT_COMPRESS || encoding == DIRECT_COMPRESS_VARCHAR) {
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/datastore/page/encoding/EncodingFactory.java]: why DIRECT_COMPRESS_VARCHAR is required?? you can use DIRECT_COMPRESS for varchar

File: core/src/main/java/org/apache/carbondata/core/datastore/page/encoding/compress/DirectCompressCodec.java [hunk 6]
```diff
@@ -64,7 +64,7 @@ public ColumnPageDecoder createDecoder(ColumnPageEncoderMeta meta) {
     return new DirectDecompressor(meta);
   }
 
-  private static class DirectCompressor extends ColumnPageEncoder {
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/datastore/page/encoding/compress/DirectCompressCodec.java]: Why this change is required ??

File: core/src/main/java/org/apache/carbondata/core/datastore/page/statistics/LVLongStringStatsCollector.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  @Override
+  protected byte[] getActualValue(byte[] value) {
+    byte[] actualValue;
+    assert (value.length >= 4);
+    if (value.length == 4) {
+      assert (value[0] == 0 && value[1] == 0);
+      actualValue = new byte[0];
+    } else {
+      // todo: what does this mean?
+      // int length = (value[0] << 8) + (value[1] & 0xff);
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/datastore/page/statistics/LVLongStringStatsCollector.java]: Here first four bytes you have to consider to get the length of the data, please refer code in ShortLength collector

File: core/src/main/java/org/apache/carbondata/core/metadata/blocklet/BlockletInfo.java [hunk 8]
```diff
@@ -205,7 +205,7 @@ public void setNumberOfPages(int numberOfPages) {
     output.writeLong(dimensionOffset);
     output.writeLong(measureOffsets);
     int dsize = dimensionChunkOffsets != null ? dimensionChunkOffsets.size() : 0;
-    output.writeShort(dsize);
+    output.writeInt(dsize);
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/metadata/blocklet/BlockletInfo.java]: Why you have changed this.
This is for number of dimension columns, how it will change for varchar data type ??

File: core/src/main/java/org/apache/carbondata/core/metadata/blocklet/BlockletInfo.java [hunk 9]
```diff
@@ -268,7 +268,7 @@ private DataChunk deserializeDataChunk(byte[] bytes) throws IOException {
   @Override public void readFields(DataInput input) throws IOException {
     dimensionOffset = input.readLong();
     measureOffsets = input.readLong();
-    short dimensionChunkOffsetsSize = input.readShort();
+    int dimensionChunkOffsetsSize = input.readInt();
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/metadata/blocklet/BlockletInfo.java]: Same as above

**Comment by kumarvishal09** [comment]
@xuchuanyin Please remove all the changes which is not related to this pr

File: integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala [hunk 10]
```diff
@@ -279,7 +279,7 @@ abstract class CarbonDDLSqlParser extends AbstractCarbonSparkSQLParser {
     fields.zipWithIndex.foreach { case (field, index) =>
       field.schemaOrdinal = index
     }
-    val (dims, msrs, noDictionaryDims, sortKeyDims) = extractDimAndMsrFields(
+    val (dims, msrs, noDictionaryDims, sortKeyDims, varcharColumns) = extractDimAndMsrFields(
```
Comment (kumarvishal09) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala]: why this varcharcolumns is required as varchar columns is nothing but dimension column of string data type

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RawColumnChunkUtil.java (hunk 3 above)
Comment (xuchuanyin) [reply core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RawColumnChunkUtil.java]: OK

File: core/src/main/java/org/apache/carbondata/core/datastore/page/encoding/DefaultEncodingFactory.java (hunk 4 above)
Comment (xuchuanyin) [reply core/src/main/java/org/apache/carbondata/core/datastore/page/encoding/DefaultEncodingFactory.java]: emm, better not do this in this PR. All the parameters for *IndexCodec looks alike. Changing all of them will introduce unrelated changes.

File: core/src/main/java/org/apache/carbondata/core/datastore/page/encoding/EncodingFactory.java (hunk 5 above)
Comment (xuchuanyin) [reply core/src/main/java/org/apache/carbondata/core/datastore/page/encoding/EncodingFactory.java]: `DIRECT_COMPRESS_VARCHAR` is a special encoding to distinguish normal string and long string. It is used in CompressedDimensionChunkFileBasedReaderV3 to indicate the DimensionStoreType.

File: core/src/main/java/org/apache/carbondata/core/datastore/page/encoding/compress/DirectCompressCodec.java (hunk 6 above)
Comment (xuchuanyin) [reply core/src/main/java/org/apache/carbondata/core/datastore/page/encoding/compress/DirectCompressCodec.java]: Yeah, it is required because in the method `getEncodingList`, we want to use member `datatype` from the outside class. If it is static inner class, we cannot access that member.

File: core/src/main/java/org/apache/carbondata/core/metadata/blocklet/BlockletInfo.java (hunk 8 above)
Comment (xuchuanyin) [reply core/src/main/java/org/apache/carbondata/core/metadata/blocklet/BlockletInfo.java]: OK~

File: core/src/main/java/org/apache/carbondata/core/metadata/blocklet/BlockletInfo.java (hunk 9 above)
Comment (xuchuanyin) [reply core/src/main/java/org/apache/carbondata/core/metadata/blocklet/BlockletInfo.java]: OK~

File: integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala (hunk 10 above)
Comment (xuchuanyin) [reply integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala]: Just like the other results of `extractDimAndMsrFields`, we validate and get the sort_column, dictionaries and the varcharColumns(longStringColumns).

For the varcharColumns, we change their datatype from string to varchar later.

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java (hunk 1 above)
Comment (xuchuanyin) [reply core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: As I mentioned in another PR, better not to introduce this limits.
-1 means that the parser can parse infinity characters.

File: core/src/main/java/org/apache/carbondata/core/datastore/page/VarLengthColumnPageBase.java (hunk 2 above)
Comment (xuchuanyin) [reply core/src/main/java/org/apache/carbondata/core/datastore/page/VarLengthColumnPageBase.java]: I come across a new idea:
During parsing/converting, we can calculate #numberOfRowsPerPage * #currentCharacterLength, if it is larger than 2GB, dataload will fail.
Notice that the #numberOfRowsPerPage is specified by user through configurations.

If this is OK, I'll implementation in the future not this PR.

File: core/src/main/java/org/apache/carbondata/core/datastore/page/statistics/LVLongStringStatsCollector.java (hunk 7 above)
Comment (xuchuanyin) [reply core/src/main/java/org/apache/carbondata/core/datastore/page/statistics/LVLongStringStatsCollector.java]: yeah, I find a more readable way to fix it.

**Comment by xuchuanyin** [comment]
Rebased with the latest master branch.

The second commit is to fix the review comments.

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5230/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6396/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5342/

File: core/src/main/java/org/apache/carbondata/core/datastore/page/encoding/EncodingFactory.java (hunk 5 above)
Comment (kumarvishal09) [reply core/src/main/java/org/apache/carbondata/core/datastore/page/encoding/EncodingFactory.java]: @xuchuanyin I think in your last PR it was VARCHAR...

**Comment by xuchuanyin** [comment]
retest this please

**Comment by xuchuanyin** [comment]
retest it please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5239/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6405/

**Comment by kumarvishal09** [comment]
@xuchuanyin It is difficult for the user to configure number of rows in a page as it depends on data, its better to handle internally based on size of data divide the page in multiple pages in Page Producer if it crossing 2gb. Please raise a JIRA and handle this in different PR. and Complex type is also not handled please raise JIRA for the same...I will merge this PR

**Comment by kumarvishal09** [comment]
LGTM

**Comment by xuchuanyin** [comment]
Currently the long_string_columns does not support complex datatypes. Carbondata will support it later in CARBONDATA-2620

**Comment by xuchuanyin** [comment]
@kumarvishal09 
CARBONDATA-2620 and CARBONDATA-2618 were raised for these problems
